Method no wrapper

AGaliciaMartinez opened the pull request:
This is PR includes the main class Network along with the specialisations that handle the conversion between data types and matmul. It also includes a testing function to allow comparison between networks. Although this function is not in its final form, I will leave it as it is for now. 

A few differences with respect to QuOperator:

`Network` class is very similar to `QuOperator` in its logic. The main differences are:
 - We keep track of nodes. `QuOperator` navigates the graph starting from the in_edges and out_edges instead. I believe keeping track of the node reference can be useful, specially for larger tensor networks as it will significantly improve the speed on those cases. In any case, it is straightforward to keep track of nodes in most of the linear algebra operations.
- `ref_nodes` has been removed in favour of just `nodes`.
- `ignore_edges` attribute has been removed to simplify the implementation: I am not sure of the use case for this attribute but for the moment I decided to remove it just to simplify tests and `Network`.

Methods that work (if the have a [ ] is because there is some work still to do):
- `dims`
- `_check_*`: these checks are a little bit more exhausting that the ones in the original implementation of `QuOperator`.
- `_fast_constructor`
- `conj`, `transpose` and `adjoint`. These follow same logic as `QuOperator`. 
- `contract`: Same logic as `QuOperator`. It does not handle the Copy (identity) nodes yet. I will add them this feature in another PR.
- `to_array`: replaces `eval` in `QuOperator`.
- [x] `from_2d_array`: provisional. It replaces `from_tensor`. I am still not sure with the logic of this one. At this moment it handles 2D, 1d and scalar arrays so at least the name has to change.
- [x] `__matmul__`: The logic changes a bit as we relax the condition of equal dims to allow multiplication of different networks. They have to be compatible and this is checked with `_match_edges_by_split`. For example: we allow matmul between networks of dimension  `[[4],[4]]` and `[[2,2],[2,2]]`. Some more exaustive tests may be needed here.
- [x] `_match_edges_by_split`. This is the function that takes care of reshaping nodes in network to allow matmul. This funciton still requires some more careful testing but should work.
- [x] `tensor`: Not finished yet but I would like to add it in this PR.

Methods that will not be included in this PR:
- `partial_trace`
- `trace`
- `norm`
- `as_quoperator`: I still need google/TensorNetwork#951 to be merged.

A few other missing things:
- [ ] License
- [x] decide what to do with the `typing` module. Some functions have inherited the typing from `QuOperator`. Consder including it for every function or remove it.
- [x] Black/PEP8

AGaliciaMartinez committed:
added method with no wrapper.

AGaliciaMartinez committed:
ADded init files.

AGaliciaMartinez committed:
Added convert.

AGaliciaMartinez committed:
Added quoperator.

AGaliciaMartinez committed:
Working version.

AGaliciaMartinez committed:
Updated notebook.

AGaliciaMartinez committed:
Update.

AGaliciaMartinez committed:
Merge branch 'main' into method_no_wrapper

AGaliciaMartinez committed:
Deleted some files.

AGaliciaMartinez committed:
Deleted more pycache files.

AGaliciaMartinez committed:
Removed (hopefully) all pycache files.

AGaliciaMartinez committed:
Merge branch 'main' of github.com:qutip/qutip-tensornetwork into method_no_wrapper

AGaliciaMartinez committed:
Another unnecessary file.

AGaliciaMartinez committed:
Added tests and polished code.

AGaliciaMartinez committed:
Removed notebook from data folder.

AGaliciaMartinez committed:
Removed quantum.py.

AGaliciaMartinez commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,576 @@
+        # dynamically searching for them when necessary. This is because
+        # searching all nodes in a large graph can be quite expensive while
+        # keeping track of them with network operations is straightforward.
+        self.nodes = set(nodes) if nodes else tn.reachable(self.in_edges
+                                                         + self.out_edges)
```
I am not sure how to store nodes. They need to be unique which is why in TensorNetwork they mostly use sets. However, this means that adding a new node will involve checking all of them to ensure the new node is unique. In most cases this is unnecessary. For instance, in matmul we copy both input networks so we now that the nodes are unique. I am thinking to change this into a list. Any thought @hodgestar, @nwlambert ?

hodgestar replied:
I think a set is fine for now. Checking "x in some_set" is very fast in Python (similar to "x in some_dict").

AGaliciaMartinez commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,576 @@
+        # I may need extra check such as for backends etc in the future
+
+        # TODO: should this be a porperty obtained from dims?
+        super().__init__(shape=(np.prod(self.dims[0]), np.prod(self.dims[1])))
```
At this moment the shape is an attribute. Given that the main objects are `in_edges` and `out_edges`, I wonder if we should define the `shape` as a property derived from those.

AGaliciaMartinez replied:
I decided to change it to an attribute.

AGaliciaMartinez committed:
Add tensornetwork as dependency.

AGaliciaMartinez committed:
Removed quantum as dependency.

AGaliciaMartinez committed:
Comment a final parethesis I forgot to comment before.

AGaliciaMartinez committed:
Included more tests for matmul and tensor.

A few bugs were also fixed.

AGaliciaMartinez committed:
Empoyed from_2d_array when possible.

There where several places in the code where I was not using the
method from_2d_array and instead I rewrote it in place.

AGaliciaMartinez committed:
Use from_2d_array to reduce code lines.

AGaliciaMartinez committed:
Updated shape to be a property.

I also updated the docstrings.

AGaliciaMartinez committed:
Removed the use of typing.

hodgestar commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,627 @@
+
+        return Network._fast_constructor(out_edges, in_edges, nodes)
+
+    def transopose(self):
```
```suggestion
    def transpose(self):
```

AGaliciaMartinez committed:
Added example notebook.

AGaliciaMartinez committed:
Added matplotlib to tests requirements.

This was needed to avoid CI complaining from the qutip warning that is
raised when matplotlib is not installed. Not the ideal solution but it
is the simplest one.

hodgestar commented on src/qutip_tensornetwork/core/data/matmul.py:
```diff
@@ -0,0 +1,30 @@
+    # This part of the operation wont be added until we think about what to do
+    # with addition.
+    if out is not None:
+        raise NotImplementedError()
```
```suggestion
        raise NotImplementedError("The addition of Network data layer instances is not yet implemented")
```

hodgestar commented on src/qutip_tensornetwork/core/data/matmul.py:
```diff
@@ -0,0 +1,30 @@
+    if scale != 1:
+        out._nodes = out._nodes | tn.Node(scale)
+
+    return left@right
```
```suggestion
    return left @ right
```

hodgestar commented on src/qutip_tensornetwork/testing.py:
```diff
@@ -0,0 +1,100 @@
+    ------
+    AssertionError
+        If actual and desired do not represent the same map.
+    ValueError
+        If more than one node in the network is close to each other.
```
```suggestion
```
This function is named `assert_network_close` so it should only raise AssertionErrors. It's okay if some of the helper functions it calls raise other exceptions, but this function should probably catch those and convert them into AssertionErrors.

AGaliciaMartinez replied:
Currently the function raises `ValueError` because is not able to deal with networks that have more than one similar node. I will change the `ValueError` to `NotImplementedError` to make it clear that this is a missing feature but we should change to use the package you mentioned for this task.

hodgestar commented on src/qutip_tensornetwork/testing.py:
```diff
@@ -0,0 +1,100 @@
+                # and connected to same index.
+                assert_no_edge_between(node_dict[edge.node1], node_dict[edge.node2])
+
+def assert_no_edge_between(node1, node2):
```
```suggestion
def assert_edge_between(node1, node2):
```
This function should be called `assert_edge_between` because it asserts that there *is* an edge shared between node1 and node2?

AGaliciaMartinez replied:
No, it only asserts if the for loop ends, which happens only when there is no edge between the given nodes. I changed it in an attempt to make it more clear.

hodgestar commented on src/qutip_tensornetwork/testing.py:
```diff
@@ -0,0 +1,100 @@
+import numpy as np
+def assert_network_close(actual, desired, atol=0, rtol=1e-7):
```
Not necessarily for this PR, but graph isomorphism is in general a complex problem, so I recommend that we hit it with an appropriate hammer -- e.g. the networkx library -- in testing.

Rough plan:

* Add networkx as a testing dependency.
* Convert actual and desired to networkx graphs (maybe this is the one we want https://networkx.org/documentation/stable/reference/classes/multidigraph.html ?)
* Use is_isomorphic with custom node_match and edge_match arguments to check for isomorphism (see https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.isomorphism.is_isomorphic.html#networkx.algorithms.isomorphism.is_isomorphic).
* Assert that the graphs are isomorphic.

Perhaps such an equality check is worth exposing publicly too, but let's try it out in tests first.

AGaliciaMartinez replied:
I agree that we should relly in other packages for testing isomorphism. I am not sure if networkx is the right tool though as it seems to be quite slow compared to other packages: https://graph-tool.skewed.de/performance.  I must admit that have not looked too much into the other packages either so it may be possible that they are not flexible enough. Lets keep this for another PR although I will add an issue to remember this needs to be changed.

hodgestar replied:
I think the page you linked to says why networkx is a good fit for our case -- our graphs will be comparatively small and its pure Python. We should probably check that the isomorphism is fast for reasonably sized networks though -- e.g. maybe a few thousand nodes -- if we're going to use it outside of tests. Probably calling back into custom node_match and edge_match functions from C is slow and a pain though, so not sure if graph-tool will support that or perform well on it.

hodgestar commented on tests/core/data/test_convert.py:
```diff
@@ -0,0 +1,44 @@
+from qutip import data
+from .conftest import random_complex_network, random_numpy_dense
+
+_network = random_complex_network(2)
```
Is there a good reason to create on random network here instead of just creating one in the functions that need them? Global mutable state can result in subtle bugs, so it's best avoided where possible.

hodgestar commented on tests/core/data/test_convert.py:
```diff
@@ -0,0 +1,44 @@
+    assert isinstance(qobj.data, Network)
+    np.testing.assert_allclose(qobj.full(), array)
+
+def test_network_to_dense():
```
Perhaps we should test a few different shapes here too, since it's fairly easy to add and might catch problems with edge cases later.

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+    node = random_node((2,))
+    network = Network([], node[0:], copy=False)
+    assert len(network.nodes) == 1
+    assert network.nodes.pop() is node
```
```suggestion
    assert network.nodes[0] is node
```
Modify the `.nodes` while checking them seems a bit odd, so let's just do `.nodes[0]` instead if we can.

AGaliciaMartinez replied:
unfortunately sets are not subscriptable. This is another reason I was considering to store nodes as lists.

AGaliciaMartinez replied:
I removed the pop as much as I could but I left it for those cases where I thought changing it for `list(network.nodes)[0]` was not worth.

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+    node2 = random_node((2,))
+    node[0] ^ node2[0]
+
+    with pytest.raises(ValueError):
```
Could we also test that these raise suitable error messages?

AGaliciaMartinez replied:
I am actually againts testingt the error message itself. I do see why testing the error type raised is important. Including the error type in the tests ensures that in the future this same function will continue raising the same error. This is important because a user may catch this specific error as part of its program's logic and run into problems if the error is not `ValueError`.  However, I think the user is not supposed to use the string from the error as part of its program logic. It is meant to be a clarification of what happened and how to use the class. In this sense, I see the error message fulfilling the same role as the documentation, which is not included in the tests either.

hodgestar replied:
Error messages are not documentation though -- they're generated by the program and so can have bugs that are not immediately obvious. E.g. I've often had cases where `raise ValueError(f"{x} must be less than {y}")` should have been `raise ValueError(f"{y} must be less than {x}")` or the error message is just nonsense or not easily readable and tests are the only way to catch those.

Yes, code should not generally depend on the error string itself, but it's fine for tests to depend on the error string if the string is generated by the code base being tested. While code shouldn't depend on the error string, humans do when they're debugging.

There are also other cases where serious bugs can result from error messages -- e.g. `raise ValueError(f"password {p} is invalid")` which is a common security hole. Luckily we don't encounter these kinds of cases too often in QuTiP.

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+        network = Network([node[0]], [node[1]], copy=False)
+
+def test_scalar_nodes():
+    """It is possible to include scalar nodes as long as they are not the only
+    nodes.
```
This comment is a bit confusing. It says "as long as they are not the only nodes" but test_scalar_network successfully builds a network with only a scalar node. Perhaps the wording just needs to change a bit?

AGaliciaMartinez replied:
`test_scalar_network` builds the network with an scalar node, `node_scalar`, and a non-scalar node, `node`.

AGaliciaMartinez replied:
At first I was not going to allow the creation of scalar networks but now I do. This test should be removed.

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+    network = Network([], [], nodes=[node])
+    np.testing.assert_allclose(node.tensor, network.to_array())
+
+def test_empty_rises():
```
```suggestion
def test_empty_raises():
```

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+    node = network.nodes.pop()
+    copied_node = copied_network.nodes.pop()
+    assert node is not copied_node
+    assert node.tensor is copied_node.tensor
```
We should assert that the copied networks are equal, even if it's just with `network.to_array()` and `copied_network.to_array()` for now.

AGaliciaMartinez replied:
I can definitely include that assertion but I do not think it is necessary here. Most of the networks in this test file have a single node and that is to facilitate these kind of comparisons. For instance, here, by doing `assert node.tensor is copied_node.tensor` we already check that both the copied and original nodes have the exact same information. In fact, they contain the _same_ tensor which means they save in memory when copying the nodes.

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+    node = random_node((2,3))
+    network = Network([node[0]], [node[1]])
+
+    assert network.shape[0] == network.transopose().shape[1]
+    assert network.shape[1] == network.transopose().shape[0]
+    assert network.dims[0] == network.transopose().dims[1]
+    assert network.dims[1] == network.transopose().dims[0]
+
+    np.testing.assert_almost_equal(node.tensor,
+                                   network.transopose().nodes.pop().tensor)
```
```suggestion
    assert network.shape[0] == network.transpose().shape[1]
    assert network.shape[1] == network.transpose().shape[0]
    assert network.dims[0] == network.transpose().dims[1]
    assert network.dims[1] == network.transpose().dims[0]

    np.testing.assert_almost_equal(node.tensor,
                                   network.transpose().nodes.pop().tensor)
```

AGaliciaMartinez replied:
Sorry for this one.

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+    np.testing.assert_almost_equal(node.tensor,
+                                   network.transopose().nodes.pop().tensor)
+
+class TestContract():
```
I like the use of the `class TestXXX` here to group things, but I think we should make groups consistently -- so definitely we should add a `class TestNetwork` (for tests for methods of the Network class) and then one group for each specialization. Tests outside that can either be loose or grouped depending on what seems to make sense.

AGaliciaMartinez replied:
All the methods here are for the `Network` class so I would rather not add an extra class layer for Network. Also, some classes only have one tests so I prefer not to surround them with a class. I can group the methods that have multiple tests each in one class though.

hodgestar replied:
Happy for you to make the call on how to group the tests nicely.

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+    node2 = random_node((2, 2, 2))
+    node1[0] ^ node2[2]
+
+    network = Network([node2[0]], [node2[1]], [node1, node2])
```
We should also just create this network within the individual tests if we can.

AGaliciaMartinez replied:
I included the node and network creation inside the methods.

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+                             ([2,10], [2,2,5]),
+                             ([2], [2]),
+                            ])
+    def test_comatible(self, dim_in, dim_out):
```
```suggestion
    def test_compatible(self, dim_in, dim_out):
```

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+        right_net = Network(right[:], [])
+        left_net = Network([], left[:])
+
+        result = left_net@right_net
```
```suggestion
        result = left_net @ right_net
```

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+        # Desired
+        left = left.tensor.reshape((1, np.prod(dim_out)))
+        right = right.tensor.reshape((np.prod(dim_in), 1))
+        desired = left@right
```
```suggestion
        desired = left @ right
```

AGaliciaMartinez replied:
I will run black after committing all the suggestions.

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+
+    def test_graph_structure(self):
+        """The operation tested here can be respresented as a graph in the
+        forllowing way:
```
```suggestion
        following way:
```

hodgestar commented on tests/core/data/test_network.py:
```diff
@@ -0,0 +1,334 @@
+    array = np.random.random(shape)
+    network = Network.from_2d_array(array)
+
+    np.testing.assert_allclose(array, network.to_array().reshape(shape))
```
Why do we need to call `.reshape(shape)` here? I would expect the shape of the input array to be retained?

AGaliciaMartinez replied:
`to_array` is a method required from `Data`. I believe it is meant to return always a 2D array. This assumption is mostly employed in the `testing_mathematics` although I realized no explicit documentation is added to this method in `Data`.

AGaliciaMartinez replied:
The assumption that to_array always returns a 2D array is also employed for the scipy integrators.

hodgestar commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,627 @@
+    dimensions are connected with edges to other arrays. An edge represents the
+    contraction that needs to be carried out between two nodes. The contraction
+    of nodes is done in a `lazy` way, which means that is only carried out when
+    necessary or explicitly required with the contract method.
```
```suggestion
    necessary or explicitly requested with the contract method.
```

hodgestar commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,627 @@
+        The network is checked for consistency: All dangling edges must either be
+        in `out_edges`, `in_edges`, or `ignore_edges`.
+
+        Parameters
```
It's more common and useful to document the parameters for `__init__` in the class docstring (since from user code one calls `MyClass(x, y, z)` and not `MyClass.__init__(...)`.

hodgestar commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,627 @@
+        operator.
+
+        The network is checked for consistency: All dangling edges must either be
+        in `out_edges`, `in_edges`, or `ignore_edges`.
```
```suggestion
        in `out_edges` or `in_edges`.
```

hodgestar commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,627 @@
+        return [out_space, in_space]
+
+    def _check_in_out_are_dangling(self):
+        for (i, e) in enumerate(self.out_edges):
+          if not e.is_dangling():
+            raise ValueError("output edge {} is not dangling!".format(i))
+        for (i, e) in enumerate(self.in_edges):
+          if not e.is_dangling():
+            raise ValueError("input edge {} is not dangling!".format(i))
```
```suggestion
        for (i, e) in enumerate(self.out_edges):
            if not e.is_dangling():
                raise ValueError("output edge {} is not dangling!".format(i))
        for (i, e) in enumerate(self.in_edges):
            if not e.is_dangling():
                raise ValueError("input edge {} is not dangling!".format(i))
```

hodgestar commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,627 @@
+
+    def _check_edges_unique(self):
+        """Check that in_edges and out_edges are unique."""
+        if (len(set(self.in_edges)) != len(self.in_edges) or
```
Should `in_edges` and `out_edges` not be unique when combined and not just separately? I.e should `len(set(in_edges + out_edges)) == len(in_edges) + len(out_edges)`?

AGaliciaMartinez replied:
Indeed, good catch. We do not want someone including the same edge to both in_edges and out_edges

hodgestar commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,627 @@
+            node = tn.Node(array)
+            return Network(node[0:1], node[1:])
+
+    def as_quoperator(self):
+        # TODO: This can not be implemented unless tensornetwork includes the 
+        # __init__ file for the quantum module.
+        raise NotImplementedError()
```
Let's just delete this for now, I think -- I can't see a reason for us to make a quoperator right now.

AGaliciaMartinez replied:
I agree.

hodgestar commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,627 @@
+
+        # connect edges to create network for the result
+        for (e_in, e_out) in zip(in_edges, out_edges):
+          _ = e_in ^ e_out
```
```suggestion
            _ = e_in ^ e_out
```

hodgestar reviewed (changes requested):
Sorry this took awhile to get to.  I think this is a solid start. Let's fix the small things and merge it and we can sort out all the other pieces in follow up PRs.

Other comments:

* I think we should remove `src/qutip_tensornetwork/file.txt`?
* I think `version.py` should not be committed?
* I think it's worthwhile running black over the code in a separate PR (or even just at the end of this PR) in order to sort out the formatting inherited from the quoperator files.

AGaliciaMartinez committed:
Update src/qutip_tensornetwork/core/data/network.py

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

AGaliciaMartinez committed:
Update src/qutip_tensornetwork/core/data/matmul.py

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

AGaliciaMartinez commented on src/qutip_tensornetwork/core/data/tensor.py:
```diff
@@ -0,0 +1,16 @@
+    """Returns a network obtained from the tensor operation between left and
+    right.
+
+    See also:
```
Reminder for myself. I think this is not properly documented.

AGaliciaMartinez commented on src/qutip_tensornetwork/testing.py:
```diff
@@ -0,0 +1,100 @@
+                assert_no_edge_between(node_dict[edge.node1], node_dict[edge.node2])
+
+def assert_no_edge_between(node1, node2):
+    """Raise an AssertionError if the given nodes do not have an edge connectin
```
```suggestion
    """Raise an AssertionError if the given nodes do not have an edge connecting
```

AGaliciaMartinez committed:
Apply suggestions from code review

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

AGaliciaMartinez committed:
Add suggestions on the testing module.

AGaliciaMartinez committed:
Implemented suggested changes in convert.

AGaliciaMartinez commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,627 @@
+        >>> net.dims
+        [[2], [2]]
+
+        One dimensional arrays are understoof as kets and cero dimensional ones
```
```suggestion
        One dimensional arrays are understood as kets and zero dimensional ones
```

AGaliciaMartinez committed:
Docstring fix.

AGaliciaMartinez committed:
PEP8

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

AGaliciaMartinez committed:
Removed unnecessary files.

AGaliciaMartinez commented on src/qutip_tensornetwork/testing.py:
```diff
@@ -0,0 +1,100 @@
+        Desired ``Network``.
+
+    atol: float, optional
+        Relative tolerance.
```
```suggestion
        Absolute tolerance.
```

AGaliciaMartinez commented on src/qutip_tensornetwork/testing.py:
```diff
@@ -0,0 +1,100 @@
+        Relative tolerance.
+
+    rtol: float, optional
+        Absolute tolerance.
```
```suggestion
        Relative tolerance.
```

AGaliciaMartinez committed:
Apply suggestions from code review

AGaliciaMartinez committed:
Added suggestions.

AGaliciaMartinez committed:
Added suggestions from code review.

AGaliciaMartinez committed:
Fixed error message.

AGaliciaMartinez committed:
Merge branch 'method_no_wrapper' of github.com:AGaliciaMartinez/qutip-tensornetwork into method_no_wrapper

AGaliciaMartinez committed:
Added suggestions from code review.

AGaliciaMartinez committed:
Fixed docstring.

AGaliciaMartinez committed:
PEP 8

AGaliciaMartinez commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,633 @@
+                    + str(out_dims)
+                )
+            else:
+                # new_shape=(e_out.dimension, e_in.dimension//e_out.dimension)
+                # new_e_in, e_in = tn.split_edge(e_in, shape=new_shape)
```
remove

AGaliciaMartinez committed:
Removed unnecessary comments.

Previous implementation of `_match_Edges_by_split` was catching the
IndexError exception and raising value Error instead. Since we added the
check for dimensions having same product, catching IndexError is no
longer necessary. The comments were referring to the previous version so
I removed them.

hodgestar commented on src/qutip_tensornetwork/core/data/convert.py:
```diff
@@ -0,0 +1,24 @@
+import tensornetwork as tn
+
+
+def from_quoperator(quoperator):
+    """Create a Network from a ``Quoperator``."""
+    return Network(quoperator.out_edges, quoperator.in_edges, quoperator.ref_nodes)
```
Do we still want to have `from_quoperator`?

hodgestar reviewed (approved):
Happy for this to be merged once the license is added, but also happy to do a last review after the license is added.

AGaliciaMartinez committed:
Added LICENSE

AGaliciaMartinez commented:
@hodgestar I have just added the same license that can be found in the quantum.py file in TensorNetwork. Could you take a look and let me know if I added it correctly. I also included a small clarification stating that changes to the original file follow the projects main license. Is this the correct approach for this case?

AGaliciaMartinez committed:
Removed creation function for quoperators as it wont be supported.

AGaliciaMartinez committed:
PEP 8

hodgestar commented on src/qutip_tensornetwork/core/data/network.py:
```diff
@@ -0,0 +1,642 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This file has been derived from the quantum/quantum.py file in the
+google/TensorNetwork repository. The above license corresponds to the one found
+in the original file. The original file has been extensively modified and these
+modifications follow the same license as the rest of the project does."""
```
```suggestion
#
# Note:
#
# This file has been derived from quantum.py in the google/TensorNetwork GitHub repository,
# https://github.com/google/TensorNetwork/blob/9b48cdd98777e4d7f26cccf29a562afe61e51da4/
# tensornetwork/quantum/quantum.py.
#
# The license above is the one from the original file. The file has been extensively modified. In
# addition to the license above, the modifications are also provided under the license in
# LICENSE.txt.
```

hodgestar replied:
I suggested changing this to a comment and I modified the text a bit to add an explicit link to the source file and to explicitly mention where the other license is.

Otherwise looks good!

AGaliciaMartinez committed:
Update src/qutip_tensornetwork/core/data/network.py

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

AGaliciaMartinez committed:
Formatting license.

AGaliciaMartinez commented:
I accepted your change so I will merger once the tests have passed. :smile:
